src/celestia/celx_internal.h

```diff
@@ -94,7 +95,160 @@ class CelxLua
+    }
+    template <typename T> T *newUserData(T a)
+    {
+        T *p = newUserData<T>();
+        *p = a;
+        return p;
+    }
+    template <typename T> T *newUserDataArray(int n)
+    {
+        return reinterpret_cast<T*>(lua_newuserdata(m_lua, sizeof(T) * n));
+    }
+    template <typename T> T *newUserDataArray(T *a, int n)
+    {
+        T *p = reinterpret_cast<T*>(lua_newuserdata(m_lua, sizeof(T) * n));
+        for (int i = 0; i < n; i++)
+            p[i] = a[i];
+        return p;
+    }
+    template <typename T> int pushClass(T a)
+    {
+        newUserData(a);
+        setClass(celxClassId(a));
+        return 1;
+    }
+
+/*    template<typename K, typename T> int push(std::map<const K, T>& a)
+    {
+        CelxLua celx(m_lua);
+        int n = a.size();
+        int i = 0;
+        T *array = celx.newUserDataArray<T>(n);
+        for (auto &it : a)
+        {
+            array[i] = it.second;
+            i++;
+        }
+        celx.push((n - 1));
+        auto lambda = [](lua_State *l)
+            {
+                CelxLua celx(l);
+
+                T *c = celx.getUserData<T>(CelxLua::localIndex(1));
+                int i = celx.getNumber(CelxLua::localIndex(2));
+                if (i < 0) return 0;
+                T ret = c[i];
+                i--;
+                celx.push(i);
+                lua_replace(l, CelxLua::localIndex(2));
+                return celx.pushClass(ret);
+            };
+        celx.push(lambda, 2);
+
+        return 1;
+    }*/
+    // FIXME: allow non-associative containers
+    template<typename T, typename C> int pushIterable(C& a)
+    {
+        CelxLua celx(m_lua);
+        int n = a.size();
+        int i = 0;
+        T *array = celx.newUserDataArray<T>(n);
+        for (auto &it : a)
+        {
+            array[i] = it.second;
+            i++;
+        }
+        celx.push((n - 1));
+        auto lambda = [](lua_State *l)
+            {
+                CelxLua celx(l);
+
+                T *c = celx.getUserData<T>(CelxLua::localIndex(1));
+                int i = celx.getNumber(CelxLua::localIndex(2));
+                if (i < 0) return 0;
+                T ret = c[i];
+                i--;
+                celx.push(i);
+                lua_replace(l, CelxLua::localIndex(2));
+                return celx.pushClass(ret);
+            };
+        celx.push(lambda, 2);
+
+        return 1;
+    }
+
+    /**** type check methods ****/
+
     bool isType(int index, int type) const;
+    bool isInteger(int n = 0) const { return lua_isinteger(m_lua, n); }
+    bool isNumber(int n = 0) const { return lua_isnumber(m_lua, n); }
+    bool isBoolean(int n = 0) const { return lua_isboolean(m_lua, n); }
+    bool isString(int n = 0) const { return lua_isstring(m_lua, n); }
+    bool isUserData(int n = 0) const { return lua_isuserdata(m_lua, n); }
+    bool isValid(int) const;
+
+    /**** get methods ****/
+
+    int getInt(int n = 0) const { return lua_tointeger(m_lua, n); }
```

Comment:
We already have `Celx::getSafe` family of funtions.

Reply:
So, here are unsafe (but faster) functions... Perhaps I wrote them for symmetry.